napari/util/keybindings.py

```diff
@@ -0,0 +1,372 @@
+
+    return components_to_comb(key, modifiers)
+
+
+UNDEFINED = object()
+
+
+def bind_key(keymap, key, func=UNDEFINED, *, overwrite=False):
+    """Bind a key combination to a keymap.
+
+    Parameters
+    ----------
+    keymap : dict of str: callable
+        Keymap to modify.
+    key : str
+        Key combination.
+    func : callable or None
+        Callable to bind to the key combination.
+        If ``None`` is passed, unbind instead.
+    overwrite : bool, keyword-only, optional
+        Whether to overwrite the key combination if it already exists.
+
+    Returns
+    -------
+    unbound : callable or None
+        Callable unbound by this operation, if any.
+
+    Notes
+    -----
+    Keys are represented in the form ``[modifier-]key``, e.g. ``a``,
+    ``Control-c``, or ``Control-Alt-Delete``.
+    Valid modifiers are Control, Alt, Shift, and Meta.
+
+    Letters will always be read as upper-case.
+    Due to the native implementation of the key system, Shift pressed in certain
+    key combinations may yield inconsistent or unexpected results.
+    Therefore, it is not recommended to use Shift with non-letter keys.
+    On OSX, Control is swapped with Meta such that pressing Command reads as
+    Control.
+
+    Special keys include Shift, Control, Alt, Meta, Up, Down, Left, Right,
+    PageUp, PageDown, Insert, Delete, Home, End, Escape, Backspace, F1,
+    F2, F3, F4, F5, F6, F7, F8, F9, F10, F11, F12, Space, Enter, and Tab
+
+    Functions take in only one argument: the parent that the function
+    was bound to. This is the viewer or layer.
+
+    By default, all functions are assumed to work on key presses only,
+    but can be denoted to work on release too by separating the function
+    into two statements with the yield keyword::
+
+        @viewer.bind_key('h')
+        def hello_world(viewer):
+            # on key press
+            viewer.status = 'hello world!'
+
+            yield
+
+            # on key release
+            viewer.status = 'goodbye world :('
+
+    """
+    if func is UNDEFINED:
+
+        def inner(func):
+            bind_key(keymap, key, func, overwrite=overwrite)
+            return func
+
+        return inner
+
+    key = normalize_key_combination(key)
+
+    if func is not None and key in keymap and not overwrite:
+        raise ValueError(
+            f'key combination {key} already used! '
+            "specify 'overwrite=True' to bypass this check"
+        )
+
+    unbound = keymap.pop(key, None)
+
+    if func is not None:
+        if not callable(func):
+            raise TypeError("'func' must be a callable")
+        keymap[key] = func
+
+    return unbound
+
+
+class KeybindingDescriptor:
+    """Descriptor which transforms ``func`` into a method with the first argument bound
+    to ``class_keymap`` or ``_keymap`` depending on if it was called
+    from the class or the instance, respectively.
+
+    Parameters
+    ----------
+    func : callable
+        Function to bind.
```

Comment:
Maybe add a usage example to the docstring? I'm not sure how this works, how to use it, or what problem this solves, based on just reading the description.